### PR TITLE
Use cached metadata instead tx to reduce disc io

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

File: pkg/model/tangle/transaction_storage.go [hunk 1]
```diff
@@ -150,19 +168,46 @@ func GetCachedTransactionOrNil(txHash hornet.Hash) *CachedTransaction {
...
 		return nil
 	}
-	return storedTx.(*hornet.Transaction)
+
+	addTrunkBranchHashesToMetadata(cachedMeta.Retain())
+
+	return &CachedMetadata{CachedObject: cachedMeta}
+}
+
+func addTrunkBranchHashesToMetadata(cachedMetadata objectstorage.CachedObject) {
```
Comment (alexsporn) [inline pkg/model/tangle/transaction_storage.go]: Rename since now it does more than that?

**Review by alexsporn — APPROVED** [review]

File: pkg/model/tangle/transaction_storage.go [hunk 2]
```diff
@@ -60,10 +66,18 @@ func (c *CachedTransaction) GetTransaction() *hornet.Transaction {
 	return c.tx.Get().(*hornet.Transaction)
 }
 
+func (c *CachedTransaction) GetCachedMetadata() *CachedMetadata {
+	return &CachedMetadata{c.metadata}
```
Comment (alexsporn) [inline pkg/model/tangle/transaction_storage.go]: this should probably retain the metadata

File: plugins/snapshot/pruning.go [hunk 3]
```diff
@@ -111,13 +111,13 @@ func pruneTransactions(txsToCheckMap map[string]struct{}) int {
 		cachedTx.Release(true) // tx -1
```
Comment (alexsporn) [inline plugins/snapshot/pruning.go]: Release at the end of the loop, or maybe do a Consume?

File: plugins/tangle/solidifier.go [hunk 4]
```diff
@@ -56,22 +56,26 @@ func TriggerSolidifier() {
...
+func markTransactionAsSolid(cachedTxMeta *tangle.CachedMetadata) {
 	// Construct the complete bundle if the tail got solid (before setting solid flag => otherwise not threadsafe)
-	if cachedTx.GetTransaction().IsTail() {
-		tangle.OnTailTransactionSolid(cachedTx.Retain()) // tx pass +1
+	if cachedTxMeta.GetMetadata().IsTail() {
+		cachedTx := tangle.GetCachedTransactionOrNil(cachedTxMeta.GetMetadata().GetTxHash())
+		if cachedTx == nil {
+			log.Panicf("markTransactionAsSolid: Transaction not found: %v", cachedTxMeta.GetMetadata().GetTxHash().Trytes())
+		}
+		tangle.OnTailTransactionSolid(cachedTx) // tx pass +1
```
Comment (alexsporn) [inline plugins/tangle/solidifier.go:66]: OnTailTransactionSolid and then tryConstructBundle can use metadata only too as far as I can see

Comment (alexsporn) [reply plugins/tangle/solidifier.go:66]: hmm... tryConstructBundle is also using CurrentIndex and LastIndex, so maybe not :(

Comment (muXxer) [reply plugins/tangle/solidifier.go:66]: No we need the Transaction indexes to construct the bundle, which are only stored in the transaction.
Was thinking about adding that to the metadata as well, but this function is only called when we receive a transaction, so we have the object in cache anyway.... and we will remove the bundle scam soom.

File: plugins/webapi/debug.go [hunk 5]
```diff
@@ -213,41 +213,41 @@ func searchEntryPoints(i interface{}, c *gin.Context, _ <-chan struct{}) {
...
 				}
 			}
 
 			return true, nil
 		},
 		// consumer
-		func(cachedTx *tangle.CachedTransaction) error { // tx +1
-			defer cachedTx.Release(true) // tx -1
+		func(cachedTxMeta *tangle.CachedMetadata) error { // meta +1
+			defer cachedTxMeta.Release(true) // meta -1
```
Comment (alexsporn) [inline plugins/webapi/debug.go]: maybe use Consume?

**Review by alexsporn — APPROVED** [review]

**Review by alexsporn — APPROVED** [review]

**Pull request merged**